src/libraries/System.Private.Xml/src/System/Xml/Core/XmlTextWriter.cs

```diff
@@ -1809,5 +1923,25 @@ private void FlushEncoders()
             _base64Encoder?.Flush();
             _flush = false;
         }
+
+        private delegate bool GrowPrimitiveBufferDelegate<in T>(T value, Span<char> destination, out int charsWritten);
+        private void WriteWithBuffer<T>(T value, GrowPrimitiveBufferDelegate<T> checkFunc)
+        {
+            ArgumentNullException.ThrowIfNull(value);
+
+            //fast path without loop
+            if (checkFunc(value, _primitivesBuffer, out int charsWritten))
+            {
+                WriteChars(_primitivesBuffer, 0, charsWritten);
+                return;
+            }
+
+            while (!checkFunc(value, _primitivesBuffer, out charsWritten))
+            {
+                _primitivesBuffer = new char[_primitivesBuffer.Length * 2];
```

Comment:
Why is a growth function needed here? Most `T` instances being uses here have small string formattings, so it seems like we could pick an initial `_primitivesBuffer` size and assert that `checkFunc` always returns true.

Assuming we have `T` whose string format can be arbitrarily large, then the cost of copying data from the intermediate buffer possibly outweighs the cost of allocation. Not to mention that this approach uses delegates to static methods, which are known to be slow to invoke.

Reply:
@eiriktsarpalis I choose 128b and removed growing functionality.